Add drop_isel

mesejo opened the pull request:
- Closes #4658 
- Use get_index(dim) in drop_sel
- [x] Tests added
- [x] Passes `pre-commit run --all-files`
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [x] New functions/methods are listed in `api.rst`

max-sixty commented on xarray/core/dataset.py:
```diff
@@ -4053,13 +4053,68 @@ def drop_sel(self, labels=None, *, errors="raise", **labels_kwargs):
+
+        Examples
+        --------
+        >>> data = np.random.randn(2, 3)
```
Generally using `np.random.randn` is discouraged because it's harder to replicate and `np.arange` make it easier to see what's happening

max-sixty commented on xarray/core/dataset.py:
```diff
@@ -4053,13 +4053,68 @@ def drop_sel(self, labels=None, *, errors="raise", **labels_kwargs):
+        --------
+        >>> data = np.random.randn(2, 3)
+        >>> labels = ["a", "b", "c"]
+        >>> ds = xr.Dataset({"A": (["x", "y"], data), "y": labels})
```
Could we add a `>>> ds` here to make it even easier to read the example?

max-sixty commented on xarray/core/dataset.py:
```diff
@@ -4053,13 +4053,68 @@ def drop_sel(self, labels=None, *, errors="raise", **labels_kwargs):
+            # is a large numpy array
+            if utils.is_scalar(pos_for_dim):
+                pos_for_dim = [pos_for_dim]
+            pos_for_dim = np.asarray(pos_for_dim).tolist()
```
Do we need to coerce to a `list` here?

max-sixty commented on xarray/core/dataset.py:
```diff
@@ -4053,13 +4053,68 @@ def drop_sel(self, labels=None, *, errors="raise", **labels_kwargs):
             labels_for_dim = np.asarray(labels_for_dim)
             try:
-                index = self.indexes[dim]
+                index = self.get_index(dim)
```
Out of interest, what's the difference between these?

mesejo replied:
The idea is that the new code will work in dimensions that have no coordinates

keewis replied:
I might be missing something, but I don't think `drop_sel` should work for dimensions without coordinates: `sel` as well as `drop_sel` work on labels, so they need the coordinates. For reference, `self.get_index(dim)` returns a `DataArray` containing the index positions, so this actually makes `drop_sel` behave like `drop_isel` for dimensions without coordinates.

dcherian replied:
`xr.DataArray(np.arange(10)).sel(dim_0=[2, 3])` This works though so it makes sense that `drop_sel` works in all the cases that `sel` does.

max-sixty commented on xarray/tests/test_dataset.py:
```diff
@@ -2410,6 +2414,27 @@ def test_drop_labels_by_keyword(self):
         with pytest.raises(ValueError):
             data.drop(dim="x", x="a")
 
+    def test_drop_labels_by_position(self):
```
These are decent. It's even better if we can have one test which can compare the whole `ds` — similar to what you did here: https://github.com/pydata/xarray/pull/4819/files#diff-2db7f6624707083db4aaab1b62eb11352200ec9d3ac1055de84877912226d7b5R2376

max-sixty reviewed (commented):
This is excellent! Thanks @mesejo 

I added a couple of minor comments.

It would be great to do this for `DataArray` too — it can be a very simple wrapper of this method.

mathause commented:
LGTM!

> It would be great to do this for DataArray too — it can be a very simple wrapper of this method.

Just to add on to this - the wrapper can be written as:

```python

class DataArray:
    ...

    def drop_isel(self, indexers=None, **indexers_kwargs):
        """Drop index positions from this dataset.
        ...
        """
        dataset = self._to_temp_dataset()
        dataset = dataset.drop_isel(indexers=indexers, **indexers_kwargs)
        return self._from_temp_dataset(dataset)

    ...

```

mesejo commented:
> This is excellent! Thanks @mesejo
> 
> I added a couple of minor comments.
> 
> It would be great to do this for `DataArray` too — it can be a very simple wrapper of this method.

Thanks for the feedback. I've addressed the issues.

max-sixty reviewed (approved):
This is great, thanks @mesejo !

mathause commented on xarray/core/dataset.py:
```diff
@@ -4053,13 +4053,76 @@ def drop_sel(self, labels=None, *, errors="raise", **labels_kwargs):
+            pos_for_dim = np.asarray(pos_for_dim)
+            index = self.get_index(dim)
+            new_index = index.delete(pos_for_dim)
+            ds = ds.loc[{dim: new_index}]
```
Would it make sense to only do this once outside of the loop?

```python
keys = dict()
for dim, pos_for_dim in indexers.items():
    ...
    keys[dim] = new_index

ds = ds.loc[keys]
```

mathause commented on xarray/tests/test_dataset.py:
```diff
@@ -2410,6 +2414,37 @@ def test_drop_labels_by_keyword(self):
+
+        actual = data.drop_isel(x=0)
+        expected = data.drop_sel(x="a")
+        assert_array_equal(actual.coords["x"], ["b"])
```
```suggestion
```
`assert_identical` should check the coords

mathause reviewed (approved):
LGTM. I added two suggestions but it's also fine to merge as is.

keewis commented on xarray/tests/test_dataset.py:
```diff
@@ -2371,8 +2371,12 @@ def test_drop_index_labels(self):
 
-        with raises_regex(ValueError, "does not have coordinate labels"):
-            data.drop_sel(y=1)
+        actual = data.drop_sel(y=[1])
+        expected = data.isel(y=[0, 2])
+        assert_identical(expected, actual)
```
as mentioned above, I think `data.drop_sel(y=...)` should not return successfully, so the check for a `ValueError` was correct. The `KeyError` below is okay, though.

keewis reviewed (commented):
I think there are a few more issues we should address in a new PR

max-sixty commented:
Thanks for spotting these @keewis, sorry I missed them.

On the `sel` fallback to `isel`, though:

> I might be missing something, but I don't think `drop_sel` should work for dimensions without coordinates: `sel` as well as `drop_sel` work on labels, so they need the coordinates

I tend to agree, but `.sel` does fall back to `.isel` for dimensions without coordinates, so the now-current code is more consistent. What are your thoughts?

mathause commented:
I suggested using `index = self.get_index(dim)` in  #4819. Mostly for consistency with `sel` & because there was no `drop_isel`. I don't have a strong opinion. I agree it's inconsistent - but it's also convenient and IMHO not surprising.

keewis commented:
I agree it's important to be consistent with `sel`. This did come as a surprise to me, though: I assumed that `sel` would only work in coordinate space, so having it switch to index space when there's no coordinate is somewhat inconsistent with that.

Also, neither `drop_sel` nor `sel` explain how they work on dimensions without coordinates in their docstrings. If we decide to keep this I think we should make sure to update the documentation.

max-sixty commented on xarray/core/dataset.py:
```diff
@@ -4053,13 +4053,78 @@ def drop_sel(self, labels=None, *, errors="raise", **labels_kwargs):
+            A        (x, y) int64 0 2 3 5
+        """
+
+        indexers = either_dict_or_kwargs(indexers, indexers_kwargs, "drop")
```
This one we should change — @mesejo would you be up for putting the one-line PR in?

mesejo replied:
Yes, I can do it. No problem, but just to be sure, `drop_sel` also has:
`labels = either_dict_or_kwargs(labels, labels_kwargs, "drop")`

Should I change it also? BTW `drop_sel` also uses `np.random.randn` in the docstrings should I change this also?

max-sixty replied:
Ah, good spot, that's a mistake, it should be ` "drop_sel")`

We're trying to move away from random values in the docstrings, but it's not urgent — if you can that would be appreciated, but correcting this & https://github.com/pydata/xarray/pull/4819#discussion_r559837727 is fine to do alone.

Thank you!
